zcash_client_sqlite/src/lib.rs

```diff
@@ -910,12 +911,37 @@ impl<P: consensus::Parameters> WalletWrite for WalletDb<rusqlite::Connection, P>
                         }
 
                         // Ensure we have a Sapling checkpoint for each checkpointed Orchard block height
+                        // We skip all checkpoints below the minimum retained checkpoint in the
+                        // Sapling tree, because branches below this height may be pruned.
                         #[cfg(feature = "orchard")]
-                        for (height, checkpoint) in &missing_sapling_checkpoints {
-                            sapling_tree
-                                .store_mut()
-                                .add_checkpoint(*height, checkpoint.clone())
+                        {
+                            let min_checkpoint_height = sapling_tree
+                                .store()
+                                .min_checkpoint_id()
                                 .map_err(ShardTreeError::Storage)?;
```

Comment:
Will this ever return `None` if called here? We insert an explicit checkpoint for `from_state`, so surely we will always have a checkpoint ID by now. I think for the intended effect, we need to call this before we make any modifications to the Sapling tree.

Reply:
This has to be called after `insert_tree`, because that is the call that triggers the pruning that then may make this insertion invalid.